File: premiere/langages/2021/exercices-fonctions/scripts/division.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
@Author: Christophe Viroulaud
@Time:   2021/07/09 16:05:38
"""


def division_rec(dividende: int, diviseur: int) -> tuple:
    if dividende < diviseur:
        return (0, dividende)
    t = division_rec(dividende-diviseur, diviseur)
    return (t[0]+1, t[1])


print(division_rec(39, 5))
print(division_rec(35, 5))


def division(dividende: int, diviseur: int) -> tuple:
    q = 0
    while dividende >= diviseur:
        q += 1
        dividende -= diviseur
    return (q, dividende)


print(division(39, 5))
print(division(35, 5))

File: premiere/langages/2021/exercices-fonctions/scripts/c/division.c
#include <stdio.h>
#include <stdlib.h>

typedef struct resultat resultat;
struct resultat {
  int q;
  int r;
};

void division(int dividende, int diviseur, resultat *res) {
  res->q = 0;
  res->r = 0;
  while (dividende >= diviseur) {
    res->q = res->q + 1;
    dividende = dividende - diviseur;
  }
  res->r = dividende;
}

void division_rec(int dividende, int diviseur, resultat *res) {
  if (dividende >= diviseur) {
    res->q++;
    division_rec(dividende - diviseur, diviseur, res);
  } else {
    res->r = dividende;
  }
}

int main(int argc, char *argv[]) {
  resultat res;
  division(39, 5, &res);
  printf("%d %d\n", res.q, res.r);

  res.q = 0;
  res.r = 0;
  division_rec(39, 5, &res);
  printf("%d %d\n", res.q, res.r);

  return 0;
}